### PR TITLE
[YUNIKORN-2553] Enable deadlock detection during unit tests

### PR DESCRIPTION
**Pull request by pbacsko**
### What is this PR for?
Enable deadlock detection when running unit tests with "make test". Original locking tests should be unaffected.

### What type of PR is it?
* [ ] - Bug Fix
* [x] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-2553

### How should this be tested?
1. Create a tiny test file eg. `pkg/scheduler/objects/mutex_test.go`:

```
package objects

import (
	"testing"
	"github.com/apache/yunikorn-core/pkg/locking"
)

func TestDeadlock(t *testing.T) {
	var mutex locking.Mutex
	mutex.Lock()
	mutex.Lock()
}
```

2. Run `make test`

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

**Comment by pbacsko** [comment]
Putting change to draft: it fails because it manages to detect problems (#840 and #841). Can go in after those PRs.

**Review by wilfred-s — APPROVED** [review]
LGTM, commit after the two changes for locking I just approved

**Comment by pbacsko** [comment]
Note: `panic()` cannot be called from `onPotentialDeadlock()` because ironically that'll create a deadlock... so let's just exit which is the simplest solution.

File: Makefile
Comment (pbacsko) [reply Makefile:205]: Yes, but in that case "make test" will not fail and that's not what we want.

Comment (chia7712) [reply Makefile:205]: Should we always make `DEADLOCK_EXIT` be true if we are running the deadlock detection? I means we don't need to have extra env variable (DEADLOCK_EXIT) to enable to exit the program in deadlocking.

Comment (craigcondit) [reply Makefile:205]: We don't want DEADLOCK_EXIT to be the default normally - users may want to enable deadlock detection on a real cluster and if the pod restarts immediately it's a lot harder to track down.

Comment (chia7712) [reply Makefile:205]: > users may want to enable deadlock detection on a real cluster and if the pod restarts immediately it's a lot harder to track down.

that makes sense to me

File: pkg/locking/locking.go
Comment (chia7712) [inline pkg/locking/locking.go:87]: this is not related to this PR, but why we don't use `log.Log(log.Diagnostics).Error`?

**Review by chia7712 — APPROVED** [review]
LGTM, and only one trivial comment left.

File: pkg/locking/locking.go
Comment (craigcondit) [reply pkg/locking/locking.go:87]: That's intentional to avoid pulling in unwanted dependencies. We want to ensure that we write this before any other subsystem is initialized, including logging (which may also use locks). It's also important enough to not risk being suppressed.

Comment (pbacsko) [reply pkg/locking/locking.go:88]: Added comment

**Review by chia7712 — APPROVED** [review]
https://github.com/apache/yunikorn-core/pull/843/commits/6fe9add7756826c639f773bdec058ebc680c1766 LGTM

**Pull request closed without merging**
